### PR TITLE
prov/verbs: Optimize inject path for single-threaded use

### PR DESCRIPTION
**Pull request by shefty**
This removes 14 instructions from the inject path when FI_THREAD_DOMAIN is selected by using pre-formatted WR structures.  Most of the changes simply move existing code to different files.  See the last commit for the real change.

### CODE REVIEW

File: prov/verbs/src/verbs_ep.c [hunk 1]
```diff
@@ -641,7 +641,10 @@ int fi_ibv_open_ep(struct fid_domain *domain, struct fi_info *info,
 		}
 
 		ep->id->context = &ep->util_ep.ep_fid.fid;
-		ep->util_ep.ep_fid.msg = &fi_ibv_msg_ep_msg_ops;
+		if (dom->util_domain.threading == FI_THREAD_DOMAIN)
```
Comment (a-ilango) [inline prov/verbs/src/verbs_ep.c]: Can this be relaxed to `dom->util_domain.threading >= FI_THREAD_ENDPOINT`? If the application can guarantee that access to TX interfaces of an endpoint would be serialized, we should be able to use the static variables without any races.

File: prov/verbs/src/verbs_msg.c [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+	static struct ibv_send_wr wr = {
+		.wr_id = VERBS_INJECT_FLAG,
+		.opcode = IBV_WR_SEND,
+		.send_flags = IBV_SEND_INLINE,
+		.sg_list = &sge,
+		.num_sge = 1,
+	};
+	struct fi_ibv_ep *ep;
+
+	ep = container_of(ep_fid, struct fi_ibv_ep, util_ep.ep_fid);
```
Comment (dmitrygladkov) [inline prov/verbs/src/verbs_msg.c]: Should it save 1-2 instruction by doing:
```
return fi_ibv_send_poll_cq_if_needed(container_of(ep_fid, struct fi_ibv_ep, util_ep.ep_fid), &wr);
```

File: prov/verbs/src/verbs_msg.c [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+
+	sge.addr = (uintptr_t) buf;
```
Comment (dmitrygladkov) [inline prov/verbs/src/verbs_msg.c]: Would it be better ?
```
static struct ibv_sge sge = {
    .addr = buf,
    .length = len,
};
```

File: prov/verbs/src/verbs_msg.c [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+
+	wr.imm_data = htonl((uint32_t) data);
+	sge.addr = (uintptr_t) buf;
+	sge.length = (uint32_t) len;
+
+	return fi_ibv_send_poll_cq_if_needed(ep, &wr);
+}
+
+
+struct fi_ops_msg fi_ibv_msg_ep_msg_ops_ts = {
```
Comment (dmitrygladkov) [inline prov/verbs/src/verbs_msg.c]: Does it mean thread-safe?
But I think that both of them are thread-safe.

**Comment by shefty** [comment]
PR updated.  I replaced the static variables with a work request stored directly as part of the endpoint.  This gives us a fast inject path for any threading level but THREAD_SAFE.  We need to decide if we want other fast work requests (inject with cq data, RMA, RMA with cq data, etc.), and, if so, if this is the way that we want to approach this.

The updated PR hasn't been tested yet.  I'll verify prior to merging if the changes look okay.

**Comment by shefty** [comment]
The cost savings using a WR attached to the EP, versus a static variable is 11 instructions (versus the original 14).

File: prov/verbs/src/verbs_msg.c (hunk 4 above)
Comment (dmitrygladkov) [reply prov/verbs/src/verbs_msg.c]: Nevermind, Got it :)

**Comment by shefty** [comment]
I will work on fast paths for the other calls in separate patches.

**Pull request merged**